R/tamper.R

```diff
@@ -0,0 +1,249 @@
+    ## turn off tracing
+    tState <- tracingState(FALSE)
+    on.exit(tracingState(tState))
+  }
+
+  calls <- sys.calls()
+  from <- 0L
+  no_calls <- length(calls)
+
+  freduce_calls <- get_pipe_stages(calls)
+  pipe_call <- get_last_pipe_call(calls)
+  freduce_calls <- freduce_calls[ freduce_calls > pipe_call ]
+  if (is.na(pipe_call)) {
+    ## tracing is handled in recover(), so put it back
+    tracingState(tState)
+    return(recover())
+  }
+
+  ## find an interesting environment to start from
+  ## options(error = tamper) produces a call to this function as an object
+  calls <- calls[- length(calls)]
+  no_calls <- no_calls - 1L
+
+  ## look for a call inserted by trace() (and don't show frames below)
+  ## this level.
+  for (i in rev(seq_len(no_calls))) {
+    calli <- calls[[i]]
+    fname <- calli[[1L]]
+    ## deparse can use more than one line
+    if (!is.na(match(deparse(fname)[1L],
+                     c("methods::.doTrace", ".doTrace")))) {
+      from <- i - 1L
+      break
+    }
+  }
+
+  ## if no trace, look for the first frame from the bottom that is not
+  ## stop, recover or tamper
+  if (from == 0L) {
+    for (i in rev(seq_len(no_calls))) {
+      calli <- calls[[i]]
+      fname <- calli[[1L]]
+      if (!is.name(fname) ||
+          is.na(match(as.character(fname),
+                      c("tamper", "recover", "stop", "Stop")))) {
+        from <- i
+        break
+      }
+    }
+  }
+
+  ## We return here if there are no frames.
+  if (from <= 0L) {
+    cat(gettext("No suitable frames for tamper()\n"))
+    return()
+  }
+
+  if (! interactive()) {
+    try({
+      dump.frames()
+      cat(gettext("tamper called non-interactively; frames dumped, use debugger() to view\n"))
+    })
+    return(NULL)
+
+  } else if (identical(getOption("show.error.messages"), FALSE)) {
+    ## from try(silent=TRUE)?
+    return(NULL)
+  }
+
+  ## ---------------------------------------------------------------
+  ## This is the main stuff
+
+  ## Get the pieces of the pipe, convert to character for printing
+  chr_chain_parts <- chain_parts_to_chr(
+    get("chain_parts", envir = sys.frame(pipe_call))
+  )
+
+  ## Get the bad pipe stage
+  bad_stage <- get_bad_stage(freduce_calls)
+
+  ## Nice printout of pipe stages
+  chr_stages <- format_pipe_stages(chr_chain_parts, bad_stage)
+
+  ## Interactive menu
+  pipe_stack <- TRUE
+  titles <- c("\nEnter a frame number, 1 to switch mode, or 0 to exit  ",
+              "\nEnter a pipe stage number, 1 to switch mode, or 0 to exit  ")
+  other_text <- c("Show pipe stages\n", "Show full stack frames\n")
+
+  stacks <- list(calls, chr_stages)
+
+  repeat {
+
+    which <- menu(
+      c(other_text[pipe_stack + 1], stacks[[pipe_stack + 1]]),
+      title = titles[pipe_stack + 1])
+
+    if (which == 1) {
+      pipe_stack <- ! pipe_stack
+
+    } else if (which == 2) {
+      eval(substitute(browser(skipCalls = skip),
+                      list(skip = 7 - pipe_call)),
```

Comment:
It depends on how and where from `browser` is called. I copied 7 from `recover()` and assumed it would work, because I am calling `browser()` exactly the same way, but actually it does not seem to. The whole `browser(skipCalls=...)` thing seems to be broken, or maybe I just misunderstand sg.

As I understand, the idea would be that the `eval()`, `browser()`, `tamper()` stacks are not listed when you type `where` in the browser. But they do. :(